README.md
# peripheral-manager-gpu

# peripheral-manager-gpu


[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg?style=for-the-badge)](https://github.com/nuvlaedge/peripheral-manager-gpu/graphs/commit-activity)
[![GitHub issues](https://img.shields.io/github/issues/nuvlaedge/peripheral-manager-gpu?style=for-the-badge&logo=github&logoColor=white)](https://GitHub.com/nuvlaedge/peripheral-manager-gpu/issues/)
[![Docker pulls](https://img.shields.io/docker/pulls/nuvlaedge/peripheral-manager-gpu?style=for-the-badge&logo=Docker&logoColor=white)](https://cloud.docker.com/u/nuvlaedge/repository/docker/nuvlaedge/peripheral-manager-gpu)
[![Docker image size](https://img.shields.io/docker/image-size/nuvlaedge/peripheral-manager-gpu?style=for-the-badge&logo=docker&logoColor=white)](https://cloud.docker.com/u/nuvlaedge/repository/docker/nuvlaedge/peripheral-manager-gpu)


![CI Build](https://github.com/nuvlaedge/peripheral-manager-gpu/actions/workflows/main.yml/badge.svg)
![CI Release](https://github.com/nuvlaedge/peripheral-manager-gpu/actions/workflows/release.yml/badge.svg)


**This repository contains the source code for the NuvlaEdge Peripheral Manager for GPU devices - this microservice is responsible for the discovery, categorization and management of all [NuvlaEdge](https://sixsq.com/nuvlaedge) GPU peripherals.**

This microservice is an integral component of the NuvlaEdge Engine.


---

**NOTE:** this microservice is part of a loosely coupled architecture, thus when deployed by itself, it might not provide all of its functionalities. Please refer to https://github.com/nuvlaedge/deployment for a fully functional deployment

---

## Build the NuvlaEdge Peripheral Manager for GPU devices

This repository is already linked with Travis CI, so with every commit, a new Docker image is released.

There is a [POM file](pom.xml) which is responsible for handling the multi-architecture and stage-specific builds.

**If you're developing and testing locally in your own machine**, simply run `docker build .` or even deploy the microservice via the local [compose files](docker-compose.yml) to have your changes built into a new Docker image, and saved into your local filesystem.

**If you're developing in a non-master branch**, please push your changes to the respective branch, and wait for Travis CI to finish the automated build. You'll find your Docker image in the [nuvladev](https://hub.docker.com/u/nuvladev) organization in Docker hub, names as _nuvladev/peripheral-manager-gpu:\<branch\>_.

## Deploy the NuvlaEdge Peripheral Manager for GPU devices

The NuvlaEdge Peripheral Manager for GPU will only work if a [Nuvla](https://github.com/nuvla/deployment) endpoint is provided and a NuvlaEdge has been added in Nuvla.

Why? Because this microservice has been built to report directly to Nuvla. Every GPU device will be registered in Nuvla and associated with **an existing** NuvlaEdge.

### Prerequisites

 - *Docker (version 18 or higher)*
 - *Docker Compose (version 1.23.2 or higher)*
 - *Linux*

### Environment variables

|                          	|                                                                                                                                                       	|
|-------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------	|
|           NUVLAEDGE_UUID 	| (**required**) before starting the microservice, make sure you export the ID of the NuvlaEdge you've created through Nuvla: `export NUVLAEDGE_UUID=<nuvlaedge id from nuvla>` 	|
| NUVLA_ENDPOINT_INSECURE 	| if you're using an insecure Nuvla endpoint, set this to `True`: `export NUVLA_ENDPOINT_INSECURE=True`                                                	|
|          NUVLA_ENDPOINT 	| if you're not using [nuvla.io](https://nuvla.io) then set this to your Nuvla endpoint: `export NUVLA_ENDPOINT=<your endpoint>`                                      	|
| | |

### Launching the NuvlaEdge Peripheral Manager for GPU devices

Simply run `docker-compose up --build`

### If Nuvla is running on `localhost`

Simply run `docker-compose -f docker-compose.localhost.yml up --build`

## Testing

This microservice is completely automated, meaning that as long as all the proper environment variables have been correctly set and the right dependencies have been met, the respective Docker container will start by itself and automatically start registering peripherals into Nuvla, in real-time.

## Contributing

This is an open-source project, so all community contributions are more than welcome. Please read [CONTRIBUTING.md](CONTRIBUTING.md)

## Copyright

Copyright &copy; 2021, SixSq SA


CHANGELOG.md
# Changelog
## [1.3.0] - 2023-04-27
### Added
 - Nuvlaedge-common library integration
### Changed
## [1.2.0] - 2022-09-01
### Added
### Changed
 - Migration to NuvlaEdge
## [1.1.0] - 2022-07-18
### Added
### Changed
 - Use common base image for all NE components
 - Downgraded from python 3.9 to 3.8 to synchronize all components
## [1.0.1] - 2021-08-12
### Added
### Changed
 - rebase for alpine 3.12
## [1.0.0] - 2021-07-26
### Added 
 - add support for Kubernetes execution
### Changed
 - refactor code
## [0.2.1] - 2021-06-25
### Added 
 - missing "six" dependency of docker-py
### Changed
 - fix issue with containers without tags
 - fix an issue in flow() function
## [0.2.0] - 2021-01-29
### Added 
 - tolerance to failures when retrieving cuda cores
### Changed
 - fix /etc bind mount
## [0.1.1] - 2020-12-04
        ### Added
        ### Changed
                  - fix initial wait for NB bootstrap
## [0.1.0] - 2020-11-25
        ### Added 
                  - get cuda cores 
                  - minor improvements
        ### Changed
## [0.0.2] - 2020-10-02
### Added 
- ONBUILD SixSq license dump
### Changed
## [0.0.1] - 2020-04-22
### Added 
- new peripheral manager for NuvlaBox GPUs
### Changed




code/tests/test_gpu.py
import mock
import unittest

from gpu.gpu import main


class TestGpu(unittest.TestCase):

    @unittest.skip("dummy test")
    def test_main(self):
        """Dummy test."""
        main = mock.Mock()
        main.return_value = None
        assert None is main()


code/discovery.py
#!/usr/bin/env python3

# -*- coding: utf-8 -*-

"""NuvlaEdge Peripheral GPU Manager

This service provides GPU discovery for the NuvlaEdge.

It provides:
    - Nvidia Docker Runtime discovery
    - a list of the devices and libraries needed to use a GPU.
    - checks if Docker is the correct version to use --gpus.
"""

import os
import csv
import json
from shutil import which
import requests
import logging
from packaging import version

from nuvlaedge.peripherals.peripheral import Peripheral


docker_socket_file = '/var/run/docker.sock'
KUBERNETES_SERVICE_HOST = os.getenv('KUBERNETES_SERVICE_HOST')
if KUBERNETES_SERVICE_HOST:
    ORCHESTRATOR = 'kubernetes'
else:
    if os.path.exists(docker_socket_file):
        import docker
        ORCHESTRATOR = 'docker'
    else:
        ORCHESTRATOR = None


logging.basicConfig(level=logging.DEBUG)
logger: logging.Logger = logging.getLogger(__name__)

identifier = 'GPU'
image = 'nuvlaedge_cuda_core_information:{}'


def read_json(json_path):
    """
    JSON reader.
    """

    with open(json_path) as f:
        dic = json.load(f)
        return dic


def get_device_type():
    """
    Gets device type (x86_64, aarch64, arm64)
    """
    return os.uname().machine


def check_cuda():
    """
    Checks if CUDA is installed and returns the version.
    """
    cuda_version = which('nvcc')

    if cuda_version is not None:
        with open(cuda_version + '/version.txt', 'r') as f:
            v = f.readline()

            return v
    else:
        return False


def nvidia_device(devices):
    """
    Checks if any Nvidia device exist
    """
    nv_devices = []

    for device in devices:
        if device.startswith('nv'):
            nv_devices.append('/dev/{}'.format(device))

    return nv_devices


def check_cuda_installation(cuda_version):
    """
    Checks if Cuda is installed.
    """
    if 'libcuda.so' in os.listdir('/usr/lib/{}-linux-gnu'.format(cuda_version)):
        return True
    else:
        return False


def build_cuda_core_docker_cli(devices):
    """
    Creates the correct device and volume structures to be passed to docker.
    """
    cli_devices = []
    cli_volumes = {}
    libs = []

    current_devices = ['/dev/{}'.format(i) for i in os.listdir('/dev/')]

    for device in devices:

        if device in current_devices:
            cli_devices.append('{0}:{0}:rwm'.format(device))

    cuda_version = get_device_type()

    # Due to differences in the implementation of the GPUs by Nvidia, in the Jetson devices, and
    #   in the discrete graphics, there is a need for different volumes.

    if cuda_version == 'aarch64':
        libcuda = '/usr/lib/{0}-linux-gnu/'.format(cuda_version)
        etc = '/etc/'
        cli_volumes[etc] = {'bind':  etc, 'mode': 'ro'}
        libs.extend([libcuda, etc])
    else:
        libcuda = '/usr/lib/{0}-linux-gnu/libcuda.so'.format(cuda_version)
        libs.extend([libcuda])
    cuda = '/usr/local/cuda'

    libs.extend([cuda])
    cli_volumes[libcuda] = {'bind': libcuda, 'mode': 'ro'}
    cli_volumes[cuda] = {'bind': cuda, 'mode': 'ro'}

    return cli_devices, cli_volumes, libs


def get_current_image_version(client):

    peripheral_version = ''
    cuda_core_version = ''

    for container in client.containers.list():
        repo_tags = container.image.attrs.get('RepoTags')
        if not repo_tags:
            continue
        img, tag = repo_tags[0].split(':')
        if img == 'nuvlaedge/peripheral-manager-gpu':
            peripheral_version = tag
        elif img == image:
            cuda_core_version = tag
    try:
        if version.parse(peripheral_version) > version.parse(cuda_core_version):
            return peripheral_version
    except version.InvalidVersion:
        pass
    else:
        return '0.0.1'


def cuda_cores(image, devices, volumes, gpus):
    """
    Starts Cuda Core container and returns the output from the container
    """

    client = docker.from_env()

    current_version = get_current_image_version(client)
    img = image.format(current_version)

    # Build Image
    if len(client.images.list(img)) == 0 and current_version != '':
        logging.info('Build CUDA Cores Image')
        client.images.build(path='.', tag=img, dockerfile='Dockerfile.gpu')

    container_name = 'get-cuda-cores'
    container = ''
    try:
        container = client.containers.run(img,
                                          name=container_name,
                                          devices=devices,
                                          volumes=volumes,
                                          remove=True)
    except docker.errors.APIError as e:
        if '409' in str(e):
            client.api.remove_container(container_name)
            container = client.containers.run(img,
                                              name=container_name,
                                              devices=devices,
                                              volumes=volumes,
                                              remove=True)
        else:
            pass
    except Exception as e:
        logging.error(f'Unable to infer CUDA cores. Reason: {str(e)}')
        pass    # let's not stop the peripheral manager just because we can't get this property

    return str(container)


def cuda_information(output):
    """
    Gets the output from the container and returns GPU information
    """
    device_information = []
    info = [i.split(":")[1] for i in output.split('\\n')[1:-1]]
    device_information.append({'unit': 'multiprocessors', 'capacity': info[3]})
    device_information.append({'unit': 'cuda-cores', 'capacity': info[4]})
    device_information.append({'unit': info[8].split()[-1], 'capacity': info[8].split()[0]})

    return info[1], device_information


def min_docker_version():
    """
    Checks if the Docker Engine version and the Docker API version are enough to run --gpus.
    """
    docker_version = docker.from_env().version()

    for i in docker_version['Components']:

        if i['Name'] == 'Engine':

            engine_version = int(i['Version'].split('.')[0])
            api_version = float(i['Details']['ApiVersion'])

            if engine_version >= 19 and api_version >= 1.4:
                return True

    return False


def search_runtime(runtime_path, host_files_path):
    """
    Checks if Nvidia Runtime exists, and reads its files.
    """

    for i in os.listdir(runtime_path):

        if 'daemon.json' in i:
            dic = read_json(runtime_path + i)

            try:
                if 'nvidia' in dic['runtimes'].keys():
                    a = read_runtime_files(host_files_path)

                    return a
                else:
                    return None
            except KeyError:
                logging.exception("Runtimes not configured in Docker configuration")
                return None

    return None


def read_runtime_files(path):
    """
    Checks if the runtime files exist, reads them, and filters the correct information.
    """
    if os.path.isdir(path) and len(os.listdir(path)) > 0:
        filtered_lines = {'devices': [], 'libraries': []}

        for i in os.listdir(path):

            with open(path + i) as csvFile:
                reader = csv.reader(csvFile)

                for line in reader:
                    try:
                        if line[0] == 'lib':
                            filtered_lines['libraries'].append(line[1].strip())

                        elif line[0] == 'dev':
                            filtered_lines['devices'].append(line[1].strip())

                        else:
                            continue
                    except IndexError:
                        continue

        return filtered_lines
    return None


def cuda_cores_information(nv_devices, gpus):

    devices, libs, _ = build_cuda_core_docker_cli(nv_devices)
    output = cuda_cores(image, devices, libs, gpus)
    if output != '':
        try:
            name, information = cuda_information(output)
            return name, information
        except Exception as e:
            logging.error(f'Exception in cudaCoresInformation. Reason: {str(e)}')

    return None, None


def flow(**kwargs):
    runtime = search_runtime(kwargs['runtime'], kwargs['host_files_path'])

    runtime_files = {
        'available': True,
        'name': 'Graphics Processing Unit',
        'vendor': 'Nvidia',
        'classes': ['gpu'],
        'identifier': identifier,
        'interface': 'gpu',
        'additional-assets': runtime
    }

    if runtime is not None:

        # GPU is present and able to be used

        # If we are running on Docker, we might be able to fetch additional information about the CUDA cores
        if ORCHESTRATOR == 'docker':
            if min_docker_version():
                logging.info('--gpus is available in Docker...')

            else:
                logging.info('--gpus is not available in Docker, but GPU usage is available')
            name, info = cuda_cores_information(runtime['devices'], True)

            runtime_files['name'] = name
            runtime_files['resources'] = info

        logging.info(runtime_files)
        return {identifier: runtime_files}

    elif len(nvidia_device(os.listdir('/dev/'))) > 0 and check_cuda_installation(get_device_type()):

        # A GPU is present, and ready to be used, but not with --gpus
        nv_devices = nvidia_device(os.listdir('/dev/'))
        _, _, formatted_libs = build_cuda_core_docker_cli(nv_devices)

        runtime = {'devices': nv_devices, 'libraries': formatted_libs}

        # only for Docker
        if ORCHESTRATOR == 'docker':
            name, info = cuda_cores_information(runtime['devices'], True)

            runtime_files['name'] = name
            runtime_files['resources'] = info

        runtime_files['additional-assets'] = runtime

        logging.info(runtime_files)
        return {identifier: runtime_files}

    else:

        # GPU is not present or not able to be used.
        logging.info('No viable GPU available.')

        return None


def gpu_check(api_url):
    """ Checks if peripheral already exists """

    logging.info('Checking if GPU already published')

    get_gpus = requests.get(api_url + '?identifier_pattern=' + identifier)

    logging.info(get_gpus.json())

    if not get_gpus.ok or not isinstance(get_gpus.json(), list) or len(get_gpus.json()) == 0:
        logging.info('No GPU published.')
        return True

    logging.info('GPU has already been published.')
    return False


if __name__ == "__main__":

    HOST_FILES = '/etcfs/nvidia-container-runtime/host-files-for-container.d/'
    RUNTIME_PATH = '/etcfs/docker/'
    gpu_peripheral: Peripheral = Peripheral('gpu')

    gpu_peripheral.run(flow, runtime=RUNTIME_PATH, host_files_path=HOST_FILES)


code/check-cuda.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Outputs some information on CUDA-enabled devices on your computer,
including current memory usage.
It's a port of https://gist.github.com/f0k/0d6431e3faa60bffc788f8b4daa029b1
from C to Python with ctypes, so it can run without compiling anything. Note
that this is a direct translation with no attempt to make the code Pythonic.
It's meant as a general demonstration on how to obtain CUDA device information
from Python without resorting to nvidia-smi or a compiled Python extension.
Author: Jan Schlüter
"""

import sys
import ctypes


# Some constants taken from cuda.h
CUDA_SUCCESS = 0
CU_DEVICE_ATTRIBUTE_MULTIPROCESSOR_COUNT = 16
CU_DEVICE_ATTRIBUTE_MAX_THREADS_PER_MULTIPROCESSOR = 39
CU_DEVICE_ATTRIBUTE_CLOCK_RATE = 13
CU_DEVICE_ATTRIBUTE_MEMORY_CLOCK_RATE = 36


def ConvertSMVer2Cores(major, minor):
    # Returns the number of CUDA cores per multiprocessor for a given
    # Compute Capability version. There is no way to retrieve that via
    # the API, so it needs to be hard-coded.
    # See _ConvertSMVer2Cores in helper_cuda.h in NVIDIA's CUDA Samples.
    return {(1, 0): 8,    # Tesla
            (1, 1): 8,
            (1, 2): 8,
            (1, 3): 8,
            (2, 0): 32,   # Fermi
            (2, 1): 48,
            (3, 0): 192,  # Kepler
            (3, 2): 192,
            (3, 5): 192,
            (3, 7): 192,
            (5, 0): 128,  # Maxwell
            (5, 2): 128,
            (5, 3): 128,
            (6, 0): 64,   # Pascal
            (6, 1): 128,
            (6, 2): 128,
            (7, 0): 64,   # Volta
            (7, 2): 64,
            (7, 5): 64,   # Turing
            }.get((major, minor), 0)


def main():
    libnames = ('libcuda.so', 'libcuda.dylib', 'cuda.dll')
    for libname in libnames:
        try:
            cuda = ctypes.CDLL(libname)
        except OSError:
            continue
        else:
            break
    else:
        raise OSError("could not load any of: " + ' '.join(libnames))

    nGpus = ctypes.c_int()
    name = b' ' * 100
    cc_major = ctypes.c_int()
    cc_minor = ctypes.c_int()
    cores = ctypes.c_int()
    threads_per_core = ctypes.c_int()
    clockrate = ctypes.c_int()
    freeMem = ctypes.c_size_t()
    totalMem = ctypes.c_size_t()

    result = ctypes.c_int()
    device = ctypes.c_int()
    context = ctypes.c_void_p()
    error_str = ctypes.c_char_p()

    result = cuda.cuInit(0)

    if result != CUDA_SUCCESS:
        cuda.cuGetErrorString(result, ctypes.byref(error_str))
        print("cuInit failed with error code %d: %s" % (result, error_str.value.decode()))
        return 1

    result = cuda.cuDeviceGetCount(ctypes.byref(nGpus))

    if result != CUDA_SUCCESS:
        cuda.cuGetErrorString(result, ctypes.byref(error_str))
        print("cuDeviceGetCount failed with error code %d: %s" % (result, error_str.value.decode()))
        return 1

    print("Found %d device(s)." % nGpus.value)

    for i in range(nGpus.value):

        result = cuda.cuDeviceGet(ctypes.byref(device), i)

        if result != CUDA_SUCCESS:
            cuda.cuGetErrorString(result, ctypes.byref(error_str))
            print("cuDeviceGet failed with error code %d: %s" % (result, error_str.value.decode()))
            return 1

        print("Device: %d" % i)

        if cuda.cuDeviceGetName(ctypes.c_char_p(name), len(name), device) == CUDA_SUCCESS:
            print("Name: %s" % (name.split(b'\0', 1)[0].decode()))

        if cuda.cuDeviceComputeCapability(ctypes.byref(cc_major), ctypes.byref(cc_minor), device) == CUDA_SUCCESS:
            print("Compute Capability: %d.%d" % (cc_major.value, cc_minor.value))

        if cuda.cuDeviceGetAttribute(ctypes.byref(cores), CU_DEVICE_ATTRIBUTE_MULTIPROCESSOR_COUNT, device) == CUDA_SUCCESS:
            print("Multiprocessors: %d" % cores.value)
            print("CUDA Cores: %s" % (cores.value * ConvertSMVer2Cores(cc_major.value, cc_minor.value) or "unknown"))
            if cuda.cuDeviceGetAttribute(ctypes.byref(threads_per_core), CU_DEVICE_ATTRIBUTE_MAX_THREADS_PER_MULTIPROCESSOR, device) == CUDA_SUCCESS:
                print("Concurrent threads: %d" % (cores.value * threads_per_core.value))

        if cuda.cuDeviceGetAttribute(ctypes.byref(clockrate), CU_DEVICE_ATTRIBUTE_CLOCK_RATE, device) == CUDA_SUCCESS:
            print("GPU clock: %g MHz" % (clockrate.value / 1000.))

        if cuda.cuDeviceGetAttribute(ctypes.byref(clockrate), CU_DEVICE_ATTRIBUTE_MEMORY_CLOCK_RATE, device) == CUDA_SUCCESS:
            print("Memory clock: %g MHz" % (clockrate.value / 1000.))

        result = cuda.cuCtxCreate(ctypes.byref(context), 0, device)

        if result != CUDA_SUCCESS:
            cuda.cuGetErrorString(result, ctypes.byref(error_str))
            print("cuCtxCreate failed with error code %d: %s" % (result, error_str.value.decode()))

        else:

            result = cuda.cuMemGetInfo(ctypes.byref(freeMem), ctypes.byref(totalMem))

            if result == CUDA_SUCCESS:
                print("Total Memory: %ld MiB" % (totalMem.value / 1024**2))
                print("Free Memory: %ld MiB" % (freeMem.value / 1024**2))

            else:
                cuda.cuGetErrorString(result, ctypes.byref(error_str))
                print("cuMemGetInfo failed with error code %d: %s" % (result, error_str.value.decode()))

            cuda.cuCtxDetach(context)
            
    return 0


if __name__=="__main__":
    sys.exit(main())


code/gpu/gpu.py
# -*- coding: utf-8 -*-

"""NuvlaEdge Peripheral GPU Manager

This service provides GPU discovery for the NuvlaEdge.

It provides:
    - Nvidia Docker Runtime discovery
    - a list of the devices and libraries needed to use a GPU.
    - checks if Docker is the correct version to use --gpus.
"""

import os
import csv
import json
from shutil import which
import requests
import logging
from threading import Event
import time
from packaging import version

docker_socket_file = '/var/run/docker.sock'
KUBERNETES_SERVICE_HOST = os.getenv('KUBERNETES_SERVICE_HOST')
if KUBERNETES_SERVICE_HOST:
    ORCHESTRATOR = 'kubernetes'
    agent_dns_name = f'agent.{os.getenv("MY_NAMESPACE", "nuvlaedge")}'
else:
    agent_dns_name = 'agent'
    if os.path.exists(docker_socket_file):
        import docker
        ORCHESTRATOR = 'docker'
    else:
        ORCHESTRATOR = None


logging.basicConfig(level=logging.INFO)
identifier = 'GPU'
image = 'nuvlaedge_cuda_core_information:{}'


def wait_bootstrap(healthcheck_endpoint=f"http://{agent_dns_name}/api/healthcheck"):
    """ Simply waits for the NuvlaEdge to finish bootstrapping, by pinging the Agent API
    :returns
    """

    logging.info("Checking if NuvlaEdge has been initialized...")

    while True:
        try:
            r = requests.get(healthcheck_endpoint)
        except requests.exceptions.ConnectionError as e:
            logging.warning(f'Unable to establish connection with NuvlaEdge Agent: {e}. Will keep trying...')
            time.sleep(10)
            continue

        if r.ok:
            break

    logging.info('NuvlaEdge has been initialized.')
    return


def publish(url, assets):
    """
    API publishing function.
    """

    x = requests.post(url, json=assets)
    return x.json()


def readJson(jsonPath):
    """
    JSON reader.
    """

    with open(jsonPath) as f:
        dic = json.load(f)
        return dic


def getDeviceType():
    """
    Gets device type (x86_64, aarch64, arm64)
    """
    return os.uname().machine


def checkCuda():
    """
    Checks if CUDA is installed and returns the version.
    """
    version = which('nvcc')

    if version is not None:
        with open(version + '/version.txt', 'r') as f:
            v = f.readline()

            return v
    else:
        return False


def nvidiaDevice(devices):
    """
    Checks if any Nvidia device exist
    """
    nvDevices = []

    for device in devices:
        if device.startswith('nv'):
            nvDevices.append('/dev/{}'.format(device))

    return nvDevices


def checkCudaInstalation(version):
    """
    Checks if Cuda is installed.
    """
    if 'libcuda.so' in os.listdir('/usr/lib/{}-linux-gnu'.format(version)):
        return True
    else:
        return False


def buildCudaCoreDockerCLI(devices):
    """
    Creates the correct device and volume structures to be passed to docker.
    """
    cli_devices = []
    cli_volumes = {}
    libs = []

    current_devices = ['/dev/{}'.format(i) for i in os.listdir('/dev/')]

    for device in devices:

        if device in current_devices:
            cli_devices.append('{0}:{0}:rwm'.format(device))

    version = getDeviceType()

    # Due to differences in the implementation of the GPUs by Nvidia, in the Jetson devices, and
    #   in the discrete graphics, there is a need for different volumes.

    if version == 'aarch64':
        libcuda = '/usr/lib/{0}-linux-gnu/'.format(version)
        etc = '/etc/'
        cli_volumes[etc] = {'bind':  etc, 'mode': 'ro'}
        libs.extend([libcuda, etc])
    else:
        libcuda = '/usr/lib/{0}-linux-gnu/libcuda.so'.format(version)
        libs.extend([libcuda])
    cuda = '/usr/local/cuda'

    libs.extend([cuda])
    cli_volumes[libcuda] = {'bind': libcuda, 'mode': 'ro'}
    cli_volumes[cuda] = {'bind': cuda, 'mode': 'ro'}

    return cli_devices, cli_volumes, libs


def getCurrentImageVersion(client):

    peripheralVersion = ''
    cudaCoreVersion = ''

    for container in client.containers.list():
        repotags = container.image.attrs.get('RepoTags')
        if not repotags:
            continue
        img, tag = repotags[0].split(':')
        if img == 'nuvlaedge/peripheral-manager-gpu':
            peripheralVersion = tag
        elif img == image:
            cudaCoreVersion = tag

    if version.parse(peripheralVersion) > version.parse(cudaCoreVersion):
        return peripheralVersion

    else:
        return '0.0.1'


def cudaCores(image, devices, volumes, gpus):
    """
    Starts Cuda Core container and returns the output from the container
    """

    client = docker.from_env()

    currentVersion = getCurrentImageVersion(client)
    img = image.format(currentVersion)

    # Build Image
    if len(client.images.list(img)) == 0 and currentVersion != '':
        logging.info('Build CUDA Cores Image')
        client.images.build(path='.', tag=img, dockerfile='Dockerfile.gpu')

    container_name = 'get-cuda-cores'
    container = ''
    try:
        container = client.containers.run(img,
                                          name=container_name,
                                          devices=devices,
                                          volumes=volumes,
                                          remove=True)
    except docker.errors.APIError as e:
        if '409' in str(e):
            client.api.remove_container(container_name)
            container = client.containers.run(img,
                                              name=container_name,
                                              devices=devices,
                                              volumes=volumes,
                                              remove=True)
        else:
            pass
    except Exception as e:
        logging.error(f'Unable to infer CUDA cores. Reason: {str(e)}')
        pass    # let's not stop the peripheral manager just because we can't get this property

    return str(container)


def cudaInformation(output):
    """
    Gets the output from the container and returns GPU information
    """
    device_information = []
    info = [i.split(":")[1] for i in output.split('\\n')[1:-1]]
    # device_information['device-name'] = info[1]
    device_information.append({'unit': 'multiprocessors', 'capacity': info[3]})
    device_information.append({'unit': 'cuda-cores', 'capacity': info[4]})
    # device_information['gpu-clock'] = info[6]
    # device_information['memory-clock'] = info[7]
    device_information.append({'unit': info[8].split()[-1], 'capacity': info[8].split()[0]})

    return info[1], device_information


def dockerVersion():
    """
    Checks if the Docker Engine version and the Docker API version are enough to run --gpus.
    """
    version = docker.from_env().version()

    for i in version['Components']:

        if i['Name'] == 'Engine':

            engineVersion = int(i['Version'].split('.')[0])
            apiVersion = float(i['Details']['ApiVersion'])

            if engineVersion >= 19 and apiVersion >= 1.4:
                return True

    return False


def searchRuntime(runtimePath, hostFilesPath):
    """
    Checks if Nvidia Runtime exists, and reads its files.
    """

    for i in os.listdir(runtimePath):

        if 'daemon.json' in i:
            dic = readJson(runtimePath + i)

            try:
                if 'nvidia' in dic['runtimes'].keys():
                    a = readRuntimeFiles(hostFilesPath)

                    return a
                else:
                    return None
            except KeyError:
                logging.exception("Runtimes not configured in Docker configuration")
                return None

    return None


def readRuntimeFiles(path):
    """
    Checks if the runtime files exist, reads them, and filters the correct information.
    """
    if os.path.isdir(path) and len(os.listdir(path)) > 0:
        filteredLines = {'devices': [], 'libraries': []}

        for i in os.listdir(path):

            with open(path + i) as csvFile:
                reader = csv.reader(csvFile)

                for line in reader:
                    try:
                        if line[0] == 'lib':
                            filteredLines['libraries'].append(line[1].strip())

                        elif line[0] == 'dev':
                            filteredLines['devices'].append(line[1].strip())

                        else:
                            continue
                    except IndexError:
                        continue

        return filteredLines
    return None


def cudaCoresInformation(nvDevices, gpus):

    devices, libs, _ = buildCudaCoreDockerCLI(nvDevices)
    output = cudaCores(image, devices, libs, gpus)
    if output != '':
        try:
            name, information = cudaInformation(output)
            return name, information
        except Exception as e:
            logging.error(f'Exception in cudaCoresInformation. Reason: {str(e)}')

    return None, None


def flow(runtime, hostFilesPath):
    runtime = searchRuntime(runtime, hostFilesPath)

    runtimeFiles = {
        'available': True,
        'name': 'Graphics Processing Unit',
        'vendor': 'Nvidia',
        'classes': ['gpu'],
        'identifier': identifier,
        'interface': 'gpu',
        'additional-assets': runtime
    }

    if runtime is not None:

        # GPU is present and able to be used

        # If we are running on Docker, we might be able to fetch additional information about the CUDA cores
        if ORCHESTRATOR == 'docker':
            if dockerVersion():
                logging.info('--gpus is available in Docker...')

            else:
                logging.info('--gpus is not available in Docker, but GPU usage is available')
            name, info = cudaCoresInformation(runtime['devices'], True)

            runtimeFiles['name'] = name
            runtimeFiles['resources'] = info

        logging.info(runtimeFiles)
        return runtimeFiles

    elif len(nvidiaDevice(os.listdir('/dev/'))) > 0 and checkCudaInstalation(getDeviceType()):

        # A GPU is present, and ready to be used, but not with --gpus
        nvDevices = nvidiaDevice(os.listdir('/dev/'))
        _, _, formatedLibs = buildCudaCoreDockerCLI(nvDevices)

        runtime = {'devices': nvDevices, 'libraries': formatedLibs}

        # only for Docker
        if ORCHESTRATOR == 'docker':
            name, info = cudaCoresInformation(runtime['devices'], True)

            runtimeFiles['name'] = name
            runtimeFiles['resources'] = info

        runtimeFiles['additional-assets'] = runtime

        logging.info(runtimeFiles)
        return runtimeFiles

    else:

        # GPU is not present or not able to be used.
        logging.info('No viable GPU available.')

        return None


def send(url, assets):
    """ Sends POST request for registering new peripheral """

    logging.info("Sending GPU information to Nuvla")
    return publish(url, assets)


def gpuCheck(api_url):
    """ Checks if peripheral already exists """

    logging.info('Checking if GPU already published')

    get_gpus = requests.get(api_url + '?identifier_pattern=' + identifier)

    logging.info(get_gpus.json())

    if not get_gpus.ok or not isinstance(get_gpus.json(), list) or len(get_gpus.json()) == 0:
        logging.info('No GPU published.')
        return True

    logging.info('GPU has already been published.')
    return False


def main():

    API_BASE_URL = f"http://{agent_dns_name}/api"

    wait_bootstrap()

    API_URL = API_BASE_URL + "/peripheral"
    HOST_FILES = '/etcfs/nvidia-container-runtime/host-files-for-container.d/'
    RUNTIME_PATH = '/etcfs/docker/'

    e = Event()

    while True:

        gpu_peripheral = flow(RUNTIME_PATH, HOST_FILES)

        if gpu_peripheral:
            peripheral_already_registered = gpuCheck(API_URL)

            if peripheral_already_registered:
                send(API_URL, gpu_peripheral)

        e.wait(timeout=90)
